[Core] Scheduling optimization 2

rkooo567 opened the pull request:
This PR adds 2 optimization

I found the main overhead is coming from `get_max_num_running_seqs` because it needs to create a list of sequences. 

This adds 2 optimizations

1. If status is None, get_num_seqs doesn't have to create a list. 
2. If we already know num_seqs token budget is updated before calling schedule_Running, we allow to skip it from schedule_running. This will skip get_max_num_running_seqs so it improves perf

Before this PR, each iteration on A10 takes 2.6~7ms (the original scheduler logic takes about 2ms). After 1, it became 2.3~2.4 ms. After 2, it becomes 2~2.1ms

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

rkooo567 committed:
ip

rkooo567 committed:
fix one issue

rkooo567 committed:
,

rkooo567 committed:
done

rkooo567 committed:
Merge branch 'main' into scheduler-regression

rkooo567 committed:
done

rkooo567 commented on vllm/core/scheduler.py:
```diff
@@ -849,12 +858,228 @@ def _schedule_chunked_prefill(self):
             num_lookahead_slots=running_scheduled.num_lookahead_slots,
         )
 
+    def _schedule_before_regression(self) -> SchedulerOutputs:
```
ignore this func

simon-mo reviewed (approved):
LGTM given we just re-use the enable_chunking flag and removing the `before_regression_

rkooo567 committed:
done

rkooo567 commented:
this commit: Throughput: 24.59 requests/s, 12587.93 tokens/s
before regression: Throughput: 24.92 requests/s, 12757.61 tokens/s
before this commit: Throughput: 24.23 requests/s, 12404.63 tokens/s

So less than 2% diff

rkooo567 committed:
fixed

rkooo567 commented:
```
Throughput: 24.80 requests/s, 12699.84 tokens/s
```

lol I added if opposite when I removed the flag and replaced it to enable_chunk

After fixing this, the perf seems almost identical to before regrssion @simon-mo
